### PR TITLE
[macOS] Add lookupKeyForAsset to FlutterPluginRegistrar

### PR DESCRIPTION
**Pull request by zhongwuzw**
Fixes https://github.com/flutter/flutter/issues/47681

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by jmagman** [comment]
cc @cbracken

**Comment by chinmaygarde** [comment]
Ping @cbracken.

**Comment by cbracken** [comment]
Apologies; was out last week. Taking a look!

**Comment by chinmaygarde** [comment]
Gentle ping @cbracken

File: shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h:93]: We're adding enough commonality here that it probably makes sense to switch to a single header with platform #ifdef's instead of two headers, since the goal over time is to merge both headers.

@jmagman and @cbracken do you have any objection to doing that here?

File: shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm:92]: Private helpers in the macOS embedding need to be declared at the top of the file, and documented, as that's the prevailing style for this embedding.

This also needs to be named according to ObjC conventions, with the signature naming the parameter that follows it. `bundle` isn't an asset name, which is what the current naming suggests.

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:96]: Is there a reason these APIs are on FVC instead of just being public on the engine, since `engine` is public, and they are just passthroughs to the engine?

File: shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h:60]: All of these new APIs need unit tests.

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:733]: You've added two APIs, so you need at least two tests, one covering each.

File: shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h
Comment (jmagman) [reply shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h:12]: Really?  We should, though not for this PR.

File: shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h
Comment (jmagman) [reply shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h:93]: I would like to see as many headers shared as possible.  

x-ref note to self: when I tried to swap the shared modules to module imports it also required `#ifdef`s
https://github.com/flutter/flutter/issues/60025#issuecomment-726413256

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (jmagman) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:969]: Huh, the old code was missing a `dealloc` `-release`? Can you add it?

Comment (stuartmorgan) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:969]: The macOS embedding is all ARC.

File: shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h
Comment (stuartmorgan) [reply shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h:12]: The background is that the macOS embedding started life as code I was more or less the sole maintainer of, in a separate project, and I really don't like the macros. I would much rather everything be annotated so that everything is locally obvious. It also removes the conflation created by the macro between "the author thought about this, and it's non-nullable" vs. "the author forgot to think about this, so the API claims non-nullability just by default even though it's a lie, but everything still compiles because ObjC nullability is a sign not a cop". (I saw a lot of the latter in readability reviews over the years.)

But if you and @cbracken want to mass-change it, I'll just be quietly slightly sad 🙂 I do think having the engine be consistent one way or the other overall would be good.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (jmagman) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:969]: Oh oh right sorry.

**Comment by chinmaygarde** [comment]
Ping @a-wallen

**Comment by chinmaygarde** [comment]
@jmagman Are you able to review this in @a-wallen s absence?

**Comment by a-wallen** [comment]
@chinmaygarde this PR has my LGTM, but @stuartmorgan left review comments that haven't been addressed.

**Comment by chinmaygarde** [comment]
@zhongwuzw Any updates on the test for the class method?

**Comment by zhongwuzw** [comment]
> @zhongwuzw Any updates on the test for the class method?

I'll update the feedback after Chinese new year :)

**Comment by jmagman** [comment]
```
FAILED: obj/flutter/shell/platform/darwin/macos/framework/Source/flutter_framework_source.FlutterDartProject.o 
/Volumes/Work/s/w/ir/cache/goma/client/gomacc  ../../buildtools/mac-x64/clang/bin/clang++ -MD -MF obj/flutter/shell/platform/darwin/macos/framework/Source/flutter_framework_source.FlutterDartProject.o.d -DFLUTTER_FRAMEWORK -DFLUTTER_ENGINE_NO_PROTOTYPES -DUSE_OPENSSL=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D_FORTIFY_SOURCE=2 -D_LIBCPP_DISABLE_AVAILABILITY=1 -D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -DSK_TYPEFACE_FACTORY_CORETEXT -DSK_GL -DSK_METAL -DSK_SUPPORT_GPU=1 -DSK_VULKAN -DSK_CODEC_DECODES_JPEG -DSK_ENCODE_JPEG -DSK_CODEC_DECODES_PNG -DSK_ENCODE_PNG -DSK_CODEC_DECODES_WEBP -DSK_ENCODE_WEBP -DSK_HAS_WUFFS_LIBRARY -DFLUTTER_RUNTIME_MODE_DEBUG=1 -DFLUTTER_RUNTIME_MODE_PROFILE=2 -DFLUTTER_RUNTIME_MODE_RELEASE=3 -DFLUTTER_RUNTIME_MODE_JIT_RELEASE=4 -DDART_LEGACY_API=\[\[deprecated\]\] -DFLUTTER_RUNTIME_MODE=1 -DFLUTTER_JIT_RUNTIME=1 -DVULKAN_HPP_NO_EXCEPTIONS=1 -DIMPELLER_SUPPORTS_RENDERING=1 -DIMPELLER_ENABLE_METAL=1 -DIMPELLER_ENABLE_OPENGLES=1 -DIMPELLER_ENABLE_VULKAN=1 -DSK_ENABLE_DUMP_GPU -DSK_DISABLE_AAA -DSK_LEGACY_IGNORE_DRAW_VERTICES_BLEND_WITH_NO_SHADER -DSK_DISABLE_LEGACY_SHADERCONTEXT -DSK_DISABLE_LOWP_RASTER_PIPELINE -DSK_FORCE_RASTER_PIPELINE_BLITTER -DSK_METAL_WAIT_UNTIL_SCHEDULED -DSK_DISABLE_EFFECT_DESERIALIZATION -DSK_ENABLE_SKSL -DSK_ENABLE_PRECOMPILE -DSK_USE_PERFETTO -DSK_LEGACY_LAYER_BOUNDS_EXPANSION -DSK_ENABLE_API_AVAILABLE -DU_USING_ICU_NAMESPACE=0 -DU_ENABLE_DYLOAD=0 -DUSE_CHROMIUM_ICU=1 -DU_ENABLE_TRACING=1 -DU_ENABLE_RESOURCE_TRACING=0 -DU_STATIC_IMPLEMENTATION -DICU_UTIL_DATA_IMPL=ICU_UTIL_DATA_FILE -DRAPIDJSON_HAS_STDSTRING -DRAPIDJSON_HAS_CXX11_RANGE_FOR -DRAPIDJSON_HAS_CXX11_RVALUE_REFS -DRAPIDJSON_HAS_CXX11_TYPETRAITS -DRAPIDJSON_HAS_CXX11_NOEXCEPT -DVK_USE_PLATFORM_METAL_EXT -DVMA_DYNAMIC_VULKAN_FUNCTIONS=0 -DVMA_STATIC_VULKAN_FUNCTIONS=0 -DFLUTTER_API_SYMBOL_PREFIX=Embedder -DFLUTTER_NO_EXPORT -DSHELL_ENABLE_SOFTWARE -DSHELL_ENABLE_GL -DSHELL_ENABLE_VULKAN -DSHELL_ENABLE_METAL -I../.. -Igen -I../../third_party/libcxx/include -I../../third_party/libcxxabi/include -I../../build/secondary/third_party/libcxx/config -I../../third_party/vulkan_memory_allocator/include -I../../third_party/vulkan-deps/vulkan-headers/src/include -I../../flutter -Igen/flutter -I../../third_party/flatbuffers/include -I../../third_party/skia -I../../flutter/third_party -I../../third_party/dart/runtime -I../../third_party/dart/runtime/include -I../../flutter/third_party/txt/src -I../../third_party/harfbuzz/src -I../../third_party/icu/source/common -I../../third_party/icu/source/i18n -I../../third_party/rapidjson -I../../third_party/rapidjson/include -I../../third_party/angle/include -I../../flutter/third_party/accessibility -I../../flutter/shell/platform/embedder -I../../flutter/shell/platform/darwin/common/framework/Headers -isysroot /Volumes/Work/s/w/ir/cache/osx_sdk/XCode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.sdk -mmacosx-version-min=10.14.0 -fno-strict-aliasing -fstack-protector-all -arch x86_64 -fcolor-diagnostics -Wall -Wextra -Wendif-labels -Werror -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-but-set-parameter -Wno-unused-but-set-variable -Wno-implicit-int-float-conversion -Wno-c99-designator -Wno-deprecated-copy -Wno-psabi -Wno-unqualified-std-cast-call -Wno-non-c-typedef-for-linkage -Wno-range-loop-construct -Wunguarded-availability -Wno-deprecated-declarations -fvisibility=hidden -Wstring-conversion -Wnewline-eof -O2 -fno-ident -fdata-sections -ffunction-sections -g2 -Wno-newline-eof  -Werror=overriding-method-mismatch -Werror=undeclared-selector -fobjc-arc -fvisibility-inlines-hidden -fobjc-call-cxx-cdtors -std=c++17 -fno-rtti -nostdinc++ -nostdinc++ -fvisibility=hidden -fno-exceptions  -c ../../flutter/shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm -o obj/flutter/shell/platform/darwin/macos/framework/Source/flutter_framework_source.FlutterDartProject.o
../../flutter/shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm:95:12: error: property 'enableMirrors' not found on object of type 'FlutterDartProject *'
  if (self.enableMirrors) {
           ^
1 error generated.
```

**Comment by chinmaygarde** [comment]
Perhaps https://github.com/flutter/engine/commit/650db7a72fb36d2180386c97287e7dde8929d481.

**Comment by zhongwuzw** [comment]
@jmagman @cbracken Please review again :)

File: shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm
Comment (jmagman) [inline shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm:93]: I had a lot of trouble with this iOS hardcoded path when I was working on getting iOS app extensions working.  However you can make them match for now and we can address that later when we come to it.
https://github.com/flutter/engine/pull/40332/files#diff-aff27bb72cc390eb03f0b733e1fa490102e311ffac10e8e09e8245e582484220R79

File: shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h
Comment (stuartmorgan) [reply shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h]: Why is this marked as resolved? There are still two separate headers with a ton of duplication in the current state of the PR, rather than one shared header.

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (stuartmorgan) [reply shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:140]: This was marked as resolved with no comment or change.

File: shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm
Comment (stuartmorgan) [reply shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm:84]: This was marked as resolved without any change.

**Comment by zhongwuzw** [comment]
> It's not clear to me what the state of this PR is; many of my comments were closed without any changes having been made, so maybe something wasn't pushed?

@stuartmorgan I'm sorry, seems I missed changes cause of the wrong rebases, I'll check the reviews.

File: shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h
Comment (zhongwuzw) [reply shell/platform/darwin/macos/framework/Headers/FlutterDartProject.h]: Done.

File: shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm
Comment (zhongwuzw) [reply shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm]: Done.

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (zhongwuzw) [reply shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:140]: Just keep the same with iOS, and it's not pass through to the engine, the implementation belongs to `FlutterPluginRegistrar` in Flutter Engine source file.

File: shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm
Comment (zhongwuzw) [reply shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm]: Done

File: shell/platform/darwin/common/framework/Headers/FlutterDartProject.h
Comment (stuartmorgan) [reply shell/platform/darwin/common/framework/Headers/FlutterDartProject.h:12]: > I think @stuartmorgan and I may disagree on this, but I prefer `NS_ASSUME_NONNULL_BEGIN` in headers to match the way Apple recommends in their docs. https://developer.apple.com/documentation/swift/designating-nullability-in-objective-c-apis#Annotate-Regions-as-Nonnullable

I prefer to use explicit annotations, to match the way Apple recommends in their docs: https://developer.apple.com/documentation/swift/designating-nullability-in-objective-c-apis#Annotate-Nullability-of-Individual-Declarations 😉 

(More seriously, I'm not aware of any actual recommendation by Apple to use one vs the other; on that page they document both and don't seem to indicated a preferred version. They do say you can "simplify the process" the process using the macros; but the core argument from the anti-macro camp, of which I am a member, is that simplifying the *process* at the expense of the correctness of the *result* of the process is not a good tradeoff, and the problem is right in the macro name: "assume".)

> I don't want to start moving the other way in the iOS embedder since this is now in `common` and would be shared.

We are agreed on this; every other header in the directory uses the macros currently, so this header should as well; inconsistency actually makes the problem of the macros *worse* than using the macros everywhere, so if there were to be a change to the pattern for `/common/`, it should be done as a whole in its own PR, not incrementally.

File: shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm
Comment (zhongwuzw) [reply shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm]: Done

File: shell/platform/darwin/common/framework/Headers/FlutterDartProject.h
Comment (zhongwuzw) [reply shell/platform/darwin/common/framework/Headers/FlutterDartProject.h:12]: Done

File: shell/platform/darwin/common/framework/Headers/FlutterDartProject.h
Comment (zhongwuzw) [reply shell/platform/darwin/common/framework/Headers/FlutterDartProject.h]: @jmagman Seems `API_UNAVAILABLE` causes [lint error](https://ci.chromium.org/ui/p/flutter/builders/try/Mac%20iOS%20clang-tidy/7572/overview), any ideas? 
```
❌ Failures for clang-tidy on /Volumes/Work/s/w/ir/cache/builder/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm:
/Volumes/Work/s/w/ir/cache/builder/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm:241:1: error: 'FlutterDartProject' lacks a 'dealloc' instance method but must release '_dartEntrypointArguments' [clang-analyzer-osx.cocoa.Dealloc,-warnings-as-errors]
@implementation FlutterDartProject {
^
/Volumes/Work/s/w/ir/cache/builder/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm:241:1: note: 'FlutterDartProject' lacks a 'dealloc' instance method but must release '_dartEntrypointArguments'
Suppressed 2655 warnings (2643 in non-user code, 12 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
1 warning treated as error
```

Comment (jmagman) [reply shell/platform/darwin/common/framework/Headers/FlutterDartProject.h]: That's annoying. I was able to work around this by adding `@dynamic` to the implementation here

https://github.com/flutter/engine/blob/3e5f27d4812efa0b8cd4b1fc6d7c11f80fcfe56b/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm#L288

Something like:

```objc
// This property is marked unavailable on iOS in the common header.
// That doesn't seem to be enough to prevent this property from being synthesized.
// Mark dynamic to avoid warnings.
@dynamic dartEntrypointArguments;
```

Comment (zhongwuzw) [reply shell/platform/darwin/common/framework/Headers/FlutterDartProject.h]: Done

**Comment by zhongwuzw** [comment]
@jmagman @stuartmorgan Please review again :)

File: shell/platform/darwin/common/framework/Headers/FlutterDartProject.h
Comment (stuartmorgan) [inline shell/platform/darwin/common/framework/Headers/FlutterDartProject.h:18]: This TODO was about the combination that you are doing here, so this can be removed.

File: shell/platform/darwin/common/framework/Headers/FlutterDartProject.h
Comment (stuartmorgan) [inline shell/platform/darwin/common/framework/Headers/FlutterDartProject.h:43]: The implementation of this is a single line (return [this constant](https://github.com/flutter/engine/blob/d2faa10bb784eb19b3a03737d66260d12f963b5a/shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm#L13)), so it would be just as easy to make this available on macOS and avoid the extra difference.

File: shell/platform/darwin/common/framework/Source/FlutterDartProjectHelper.mm
Comment (stuartmorgan) [inline shell/platform/darwin/common/framework/Source/FlutterDartProjectHelper.mm:9]: This is a declaration comment, so should be on the declaration rather than the definition.

File: shell/platform/darwin/common/framework/Source/FlutterDartProjectHelper.mm
Comment (stuartmorgan) [inline shell/platform/darwin/common/framework/Source/FlutterDartProjectHelper.mm:41]: Same.

File: shell/platform/darwin/ios/framework/Source/FlutterDartProject_Internal.h
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/FlutterDartProject_Internal.h:16]: These should be declared in a `common/framework/source/FlutterDartProjectHelper.h`, not twice in two different headers.

File: shell/platform/darwin/common/framework/Source/FlutterDartProjectHelper.mm
Comment (stuartmorgan) [inline shell/platform/darwin/common/framework/Source/FlutterDartProjectHelper.mm:1]: Files should generally be named by what they are, rather than what they are currently used for. How about `NSBundleUtils.mm`?

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (stuartmorgan) [reply shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:840]: This should be two tests, not one test testing two different things.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

**Comment by zhongwuzw** [comment]
@stuartmorgan @jmagman Please review again, thanks :)

**Comment by zhongwuzw** [comment]
@stuartmorgan @jmagman Friendly ping, I think it's ready to review again, any review feedback?

**Comment by gspencergoog** [comment]
@stuartmorgan @jmagman Do you have time to take another look at this PR? (Desktop Triage)

**Review by jmagman — APPROVED** [review]
Sorry for the long delay, LGTM

File: shell/platform/darwin/common/BUILD.gn
Comment (stuartmorgan) [inline shell/platform/darwin/common/BUILD.gn:48]: This doesn't cause it to be in the public headers directory of the actual final framework bundle, does it?

**Review by stuartmorgan — APPROVED** [review]
LGTM (with one question), thanks!

File: shell/platform/darwin/common/BUILD.gn
Comment (zhongwuzw) [reply shell/platform/darwin/common/BUILD.gn:48]: no,it's just be shared between macos and iOS framework internal.Not affect final public headers.

**Pull request merged**
